data/domain/cardSets.tsx

```diff
@@ -0,0 +1,134 @@
+import { Card } from "./cards";
+import { NumberMap } from "../maps";
+import { ImageData } from "./imageData";
+import { CardSetBase } from "./data/CardSetRepo";
+
+export class CardSet {
+    cards: Card[];
+    displayName: string = ""; // Name to display on UI
+    cardSetName: string = ""; // Name of the set in the cards data (differ from display names sometimes)
+    baseBonusValue: number = 0;
+    bonusText: string = "";
+    setImage: string = "";
+    setID: number = 0;
+
+    constructor(public index: number, data: CardSetBase, setCards: Card[] = []) {
+        this.baseBonusValue = data.data.scaling;
+        this.bonusText = data.data.bonus;
+        this.setID = CardSet.getSetID(this.bonusText);
+        this.displayName = cardSetMap[this.setID]?.displayName;
+        this.cardSetName = cardSetMap[this.setID]?.cardSetName;
+        this.setImage = cardSetMap[this.setID]?.image;
+        this.cards = setCards?.filter(card => card.data.category == this.cardSetName && card.displayName != "New Monster?");
+    }
+
+    getImageData = (): ImageData => {
+        return {
+            location: `${this.setImage}`,
+            width: 28,
+            height: 36
+        }
+    }
+
+    // TODO : When banners are up, change the width and height to better match the images original sizes
+    getBannerImageData = (): ImageData => {
+        return {
+            location: `Cardbanner_${this.cardSetName.replaceAll(' ', '_')}`,
+            width: 28,
+            height: 36
+        }
+    }
+
+    getLevel = (): number => {
+        if(this.cards.length == 0) {
+            return 0;
+        }
+
+        const cardsTotalStars = this.getCardsTotalStars();
+
+        switch (true) {
+            case cardsTotalStars >= this.cards.length * 6: return 5;
+            case cardsTotalStars >= this.cards.length * 5: return 4;
+            case cardsTotalStars >= this.cards.length * 4: return 3;
+            case cardsTotalStars >= this.cards.length * 3: return 2;
+            case cardsTotalStars >= this.cards.length * 2: return 1;
+            default: return 0;
+        }
+    }
+
+    getCardsTotalStars = (): number => {
+        return this.cards?.reduce((sum, card) => {return sum+(card.count > 0 ? card.getStars()+1 : 0);}, 0)
+    }
+
+    getBonus = (level: number): number => {
+        return this.baseBonusValue * level;
+    }
+
+    getBonusText = (): string => {
+        return this.bonusText.replace(/{/, this.getBonus(this.getLevel()).toString());
+    }
+
+    getEmulatedBonusText = (level: number): string => {
+        return this.bonusText.replace(/{/, this.getBonus(level).toString());
+    }
```

Comment:
it's very minor but I think youcan technically change the above function to be

```ts
getBonusText = (level: number = this.getLevel()): string => {
        return this.bonusText.replace(/{/, this.getBonus(level).toString());
    }
```

And it will make this function no longer required.

Reply:
Oh indeed, didn't though about it, much cleaner, gonna address that